### PR TITLE
feat: numeric condition pt 2 (TECH-788)

### PR DESCRIPTION
**Pull request by martinkrulltott**
Implements [TECH-788](https://jira.dhis2.org/browse/TECH-788)

Continues the progress made in https://github.com/dhis2/event-reports-app/pull/875

**Requires https://github.com/dhis2/analytics/pull/1107**

Spec: [ER:C6: Condition selector](https://docs.google.com/document/d/1qrcLpVymdBpgzJyJ3IPULtotCQdg2PWnQL-AlgntMz4/edit#)

---

---

### Key features

1. Implement range-sets for numeric conditions

---

### Description

RangeSet (same as LegendSet) is stored together with the conditions in the following configuration:

```
ui: {
    ...
    conditions: {
        someDimensionId: {
             condition: 'someString' // e.g. 'GT:50:LT:75'
             legendSet: 'someLegendSetId'
         }
    }
}
```

The benefit is that the `legendSet` is always stored in connection with the `condition` string itself. The drawback is that the logic for parsing and setting these props is a bit hard to follow. But basically, for numeric conditions, the condition modal will render either "regular" conditions ('greater than' etc) with values (as seen in PR https://github.com/dhis2/event-reports-app/pull/875) or a single rangeSet/legendSet condition (note that this can't be combined with any other condition).
Hence for all calculations when it comes to numberOfConditions etc the logic is to either look at the length of the regular conditions _or_ if there's a legendSet in use. 
What also increases the complexity is that legendSets can be used without values (e.g. "Age 10y interval" without actually selecting any intervals). So for all handling of conditions, the fact that a "condition" can be either a combination of an operator (e.g. 'GT') and a value (e.g. 50) _or_ a `legendSet` with (e.g. 0-10) _or_ without a value has to be taken into account as well.

---

### TODO

- [ ] `fetchLegendSets` uses dummy data, as the API endpoint isn't ready yet
- [ ] `dimensionType` is hardcoded to numeric for now, as it's not available from the API endpoint yet either

---

### Screenshots

image

### CODE REVIEW

File: src/actions/ui.js [hunk 1]
```diff
@@ -85,11 +85,14 @@ export const acSetUiConditions = (value) => ({
...
 export const tSetUiConditionsByDimension =
-    (inputCondition, dimension) => (dispatch, getState) => {
+    (inputCondition, dimension, legendSet) => (dispatch, getState) => {
         dispatch(
             acSetUiConditions({
                 ...sGetUiConditions(getState()),
-                [dimension]: inputCondition,
+                [dimension]:
+                    inputCondition.length || legendSet
+                        ? { condition: inputCondition, legendSet }
```
[the remark this answers is not in the document]
Comment (martinkrulltott) [reply src/actions/ui.js:94]: Yeah, I tried to describe that behaviour here: 

> What also increases the complexity is that legendSets can be used without values (e.g. "Age 10y interval" without actually selecting any intervals). So for all handling of conditions, the fact that a "condition" can be either a combination of an operator (e.g. 'GT') and a value (e.g. 50) or a legendSet with (e.g. 0-10) or without a value has to be taken into account as well.

i.e. you can have either only a `legendSet` or only an `inputCondition` or both.. so `inputCondition` can be empty as long as a `legendSet` is provided (e.g. `Age 10y interval` without any selected values).

File: src/components/Dialogs/Conditions/ConditionsManager.js [hunk 2]
```diff
@@ -27,22 +47,40 @@ const ConditionsManager = ({
...
+            )
+        ) {
+            setSelectedLegendSet(null)
+        }
+    }
 
-    const setCondition = (conditionIndex, value) =>
+    const setCondition = (conditionIndex, value) => {
         setConditionsList(
             conditionsList.map((condition, index) =>
```
[the remark this answers is not in the document]
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/ConditionsManager.js:79]: Yeah, but since you need to pass it in as a value to `setConditionsList` (and can't operate on the list directly, you'd have to clone the list, change the value for the index, then pass the new list to the fn. This essentially does this but on one line. As far as looping through vs cloning goes, I suppose it'll be rare to see more than 10 conditions used at once, so from a performance perspective it should be similar?

File: src/components/Dialogs/Conditions/NumericCondition.js [hunk 3]
```diff
@@ -56,16 +227,74 @@ const NumericCondition = ({ condition, onChange, onRemove }) => {
                         label={operators[key]}
                     />
                 ))}
+                {availableLegendSets && <MenuDivider dense />}
+                {availableLegendSets && (
```
[the remark this answers is not in the document]
Comment (martinkrulltott) [reply src/components/Dialogs/Conditions/NumericCondition.js:231]: Yes normally I would, but `SingleSelectField` is very picky about its children and you can't return two components without wrapping them in fragment `<>    </>`, which messes with the `key` prop, so they need to be kept separate to both be direct children of `SingleSelectField` afaik.

File: src/components/Dialogs/Conditions/ConditionsManager.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (edoardo) [reply src/components/Dialogs/Conditions/ConditionsManager.js:79]: Yeah, I had a similar problem recently and ended up copying the array and using splice, in that case I also don't expect hundreds of elements in the list so it probably doesn't make much difference.
👍

File: src/components/Dialogs/Conditions/NumericCondition.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (edoardo) [reply src/components/Dialogs/Conditions/NumericCondition.js:231]: Ah! ok, I didn't think about the `key` prop issue 👍

**Review by edoardo — APPROVED** [review]
